components/script/script_thread.rs

```diff
@@ -1516,13 +1516,20 @@ impl ScriptThread {
             let _ = self.constellation_chan.send(ConstellationMsg::PipelineExited(id));
         }
 
-        let no_pending_loads = self.incomplete_loads.borrow().is_empty();
-        let no_remaining_contexts = self.documents.borrow().is_empty();
+        debug!("Exited pipeline {}.", id);
+    }
 
-        debug!("Exited pipeline {:?} ({}&{}).", id, no_pending_loads, no_remaining_contexts);
+    /// Handles a request to exit the script thread and shut down layout.
+    fn handle_exit_script_thread_msg(&self) {
+        debug!("Exiting script thread.");
+
+        while let Some(pipeline_id) = self.incomplete_loads.borrow().iter().next().map(|load| load.pipeline_id)
+            .or_else(|| self.documents.borrow().iter().next().map(|doc| doc.global().pipeline_id()))
```

Comment:
There is new iterator each time, but that's okay because `handle_exit_pipeline_msg` removes the entry, so we don't end up infinitely looping. We can just collect the iterators into a vector though, that is probably easier to read.